# Flight Operations — Complete Pandas Data Analysis

## Objective
Analyze the Flight Operations dataset using **Pandas** and follow a complete data analysis workflow:

1. Load and understand the dataset
2. Inspect structure, data types, missing values, and statistics
3. Select and filter relevant records
4. Sort records
5. Group and aggregate data
6. Transform data using derived columns
7. Answer meaningful business questions
8. Summarize 5–8 key observations

**Dataset:** `Day10_Flight_Operations_Dataset.csv`

## 1. Import Libraries

In [ ]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

## 2. Load the Dataset

The notebook first checks the common Google Colab location and then the current working directory. If the CSV is uploaded manually to Colab, upload `Day10_Flight_Operations_Dataset.csv` before running this cell.

In [ ]:
from pathlib import Path

possible_paths = [
    Path("/content/Flight_Operations.csv"),
    Path("/content/Day10_Flight_Operations_Dataset.csv"),
    Path("Flight_Operations.csv"),
    Path("Day10_Flight_Operations_Dataset.csv")
]

data_path = next((p for p in possible_paths if p.exists()), None)

if data_path is None:
    raise FileNotFoundError(
        "Flight_Operations.csv was not found. Upload the dataset to Google Colab and run this cell again."
    )

df = pd.read_csv(data_path)
print(f"Loaded: {data_path}")
print(f"Shape: {df.shape}")
df.head()

Loaded: /content/Day10_Flight_Operations_Dataset.csv
Shape: (180, 17)


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.00,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.00,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.00,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.00,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.00,Delayed,Cloudy,Airline Website,25,No Meal,5


## 3. Understand the Dataset

### Basic structure

In [ ]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nFirst 5 rows:")
display(df.head())

print("\nLast 5 rows:")
display(df.tail())

Rows: 180
Columns: 17

Column names:
['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination', 'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity', 'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather', 'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference', 'Passenger_Satisfaction']

Data types:
Flight_ID                  object
Flight_Date                object
Airline                    object
Origin                     object
Destination                object
Aircraft                   object
Travel_Class               object
Passengers                  int64
Seat_Capacity               int64
Average_Ticket_Price        int64
Delay_Minutes             float64
Flight_Status              object
Weather                    object
Booking_Channel            object
Avg_Baggage_Kg              int64
Meal_Preference            object
Passenger_Satisfaction      int64
dtype: object

First 5 rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.00,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.00,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.00,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.00,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.00,Delayed,Cloudy,Airline Website,25,No Meal,5



Last 5 rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.00,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.00,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.00,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.00,Delayed,Rain,Online Travel Portal,9,No Meal,2
179,FL0180,2026-01-22,Akasa Air,Delhi,Mumbai,Airbus A321,Economy,144,210,3767,75.00,Delayed,Fog,Online Travel Portal,17,No Meal,5


## 4. Check Data Quality

We check duplicate records, missing values, and unique values in categorical columns.

In [ ]:
print("Duplicate rows:", df.duplicated().sum())

print("\nMissing values:")
display(df.isna().sum().to_frame("Missing_Count"))

print("\nUnique values in categorical columns:")
for col in df.select_dtypes(include="object").columns:
    print(f"{col}: {df[col].nunique()} unique values")

Duplicate rows: 0

Missing values:


,Missing_Count
Flight_ID,0
Flight_Date,0
Airline,0
Origin,0
Destination,0
Aircraft,0
Travel_Class,0
Passengers,0
Seat_Capacity,0
Average_Ticket_Price,0



Unique values in categorical columns:
Flight_ID: 180 unique values
Flight_Date: 111 unique values
Airline: 6 unique values
Origin: 11 unique values
Destination: 8 unique values
Aircraft: 4 unique values
Travel_Class: 3 unique values
Flight_Status: 3 unique values
Weather: 5 unique values
Booking_Channel: 4 unique values
Meal_Preference: 4 unique values


## 5. Descriptive Statistics

`describe()` gives a quick statistical summary of the numerical variables.

In [ ]:
display(df.describe())

print("\nFull summary including categorical columns:")
display(df.describe(include="all").T)

,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.00,180.00,180.00,173.00,180.00,180.00
mean,122.77,197.24,"6,111.68",19.34,15.15,3.49
std,45.64,15.41,"3,615.22",29.03,5.71,1.17
min,45.00,180.00,"1,949.00",0.00,5.00,2.00
25%,78.75,186.00,"3,815.50",0.00,10.00,2.00
50%,125.00,189.00,"5,380.00",5.00,15.50,3.00
75%,162.25,210.00,"6,790.00",20.00,20.00,5.00
max,210.00,220.00,"27,832.00",110.00,25.00,5.00



Full summary including categorical columns:


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Flight_ID,180,180,FL0001,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Flight_Date,180,111,2026-02-10,6,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Airline,180,6,Vistara,37,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Origin,180,11,Delhi,43,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Destination,180,8,Delhi,51,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Aircraft,180,4,Airbus A319,60,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Travel_Class,180,3,Economy,134,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Passengers,180.00,NaN,NaN,NaN,122.77,45.64,45.00,78.75,125.00,162.25,210.00
Seat_Capacity,180.00,NaN,NaN,NaN,197.24,15.41,180.00,186.00,189.00,210.00,220.00
Average_Ticket_Price,180.00,NaN,NaN,NaN,"6,111.68","3,615.22","1,949.00","3,815.50","5,380.00","6,790.00","27,832.00"


## 6. Data Type Conversion

`Flight_Date` is converted from text to a proper datetime column so that date based analysis can be performed.

In [ ]:
df["Flight_Date"] = pd.to_datetime(df["Flight_Date"], errors="coerce")

print(df["Flight_Date"].dtype)
print("Date range:", df["Flight_Date"].min().date(), "to", df["Flight_Date"].max().date())

datetime64[ns]
Date range: 2026-01-03 to 2026-06-29


## 7. Selecting Relevant Data

### Selecting columns

In [ ]:
selected_columns = df[
    ["Flight_ID", "Airline", "Origin", "Destination",
     "Passengers", "Seat_Capacity", "Average_Ticket_Price",
     "Delay_Minutes", "Flight_Status", "Passenger_Satisfaction"]
]

display(selected_columns.head(10))

,Flight_ID,Airline,Origin,Destination,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Passenger_Satisfaction
0,FL0001,SpiceJet,Mumbai,Bengaluru,210,180,6894,50.00,Delayed,3
1,FL0002,Akasa Air,Delhi,Mumbai,70,210,4468,0.00,On Time,5
2,FL0003,Air India,Kochi,Delhi,67,220,4713,5.00,On Time,2
3,FL0004,Akasa Air,Srinagar,Delhi,153,220,3791,35.00,Delayed,2
4,FL0005,Air India,Pune,Delhi,178,180,6505,20.00,Delayed,5
5,FL0006,Akasa Air,Kochi,Delhi,123,220,5636,35.00,Delayed,2
6,FL0007,Air India Express,Delhi,Srinagar,178,210,9430,0.00,On Time,4
7,FL0008,Air India,Delhi,Chennai,157,220,14173,0.00,On Time,5
8,FL0009,Akasa Air,Delhi,Chennai,82,210,5893,75.00,Delayed,4
9,FL0010,Akasa Air,Ahmedabad,Bengaluru,125,220,2500,12.00,On Time,4


### Selecting rows with `loc` and `iloc`

In [ ]:
print("First three rows using iloc:")
display(df.iloc[:3])

print("Selected fields for the first five flights using loc:")
display(df.loc[:4, ["Flight_ID", "Airline", "Flight_Status", "Passenger_Satisfaction"]])

First three rows using iloc:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.00,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.00,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.00,On Time,Rain,Travel Agency,19,No Meal,2


Selected fields for the first five flights using loc:


,Flight_ID,Airline,Flight_Status,Passenger_Satisfaction
0,FL0001,SpiceJet,Delayed,3
1,FL0002,Akasa Air,On Time,5
2,FL0003,Air India,On Time,2
3,FL0004,Akasa Air,Delayed,2
4,FL0005,Air India,Delayed,5


## 8. Filtering Records

The following questions help identify flights with operational or customer impact characteristics.

In [ ]:
# Delayed flights
delayed_flights = df[df["Flight_Status"] == "Delayed"]
print("Number of delayed flights:", len(delayed_flights))
display(delayed_flights.head(10))

# Flights with delay above 30 minutes
delay_above_30 = df[df["Delay_Minutes"] > 30]
print("Flights delayed by more than 30 minutes:", len(delay_above_30))
display(
    delay_above_30[
        ["Flight_ID", "Airline", "Origin", "Destination", "Delay_Minutes", "Weather"]
    ].head(10)
)

# High-satisfaction flights
high_satisfaction = df[df["Passenger_Satisfaction"] >= 4]
print("Flights with satisfaction >= 4:", len(high_satisfaction))

Number of delayed flights: 59


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.00,Delayed,Storm,Online Travel Portal,23,No Meal,3
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.00,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.00,Delayed,Cloudy,Airline Website,25,No Meal,5
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.00,Delayed,Cloudy,Travel Agency,13,Vegetarian,2
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.00,Delayed,Cloudy,Mobile App,8,Vegetarian,4
16,FL0017,2026-05-24,SpiceJet,Delhi,Chennai,Airbus A321,Economy,73,220,4256,20.00,Delayed,Rain,Travel Agency,22,Non-Vegetarian,2
17,FL0018,2026-03-09,Air India Express,Jaipur,Mumbai,Airbus A321,Economy,88,180,6133,20.00,Delayed,Clear,Travel Agency,13,Vegetarian,4
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.00,Delayed,Clear,Airline Website,10,Vegetarian,2
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.00,Delayed,Clear,Mobile App,7,Vegan,4
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.00,Delayed,Storm,Airline Website,10,Vegan,3


Flights delayed by more than 30 minutes: 40


,Flight_ID,Airline,Origin,Destination,Delay_Minutes,Weather
0,FL0001,SpiceJet,Mumbai,Bengaluru,50.00,Storm
3,FL0004,Akasa Air,Srinagar,Delhi,35.00,Storm
5,FL0006,Akasa Air,Kochi,Delhi,35.00,Cloudy
8,FL0009,Akasa Air,Delhi,Chennai,75.00,Cloudy
21,FL0022,Vistara,Kochi,Delhi,75.00,Clear
26,FL0027,SpiceJet,Bengaluru,Hyderabad,110.00,Clear
29,FL0030,SpiceJet,Delhi,Srinagar,75.00,Storm
30,FL0031,Vistara,Delhi,Srinagar,50.00,Cloudy
32,FL0033,Akasa Air,Ahmedabad,Bengaluru,35.00,Cloudy
42,FL0043,Vistara,Delhi,Chennai,50.00,Clear


Flights with satisfaction >= 4: 89


## 9. Sorting Records

Sorting can reveal the flights with the highest delays, highest ticket prices, and strongest passenger satisfaction.

In [ ]:
print("Top 10 flights by delay:")
display(
    df.sort_values("Delay_Minutes", ascending=False)[
        ["Flight_ID", "Airline", "Origin", "Destination", "Delay_Minutes", "Flight_Status"]
    ].head(10)
)

print("Top 10 ticket prices:")
display(
    df.sort_values("Average_Ticket_Price", ascending=False)[
        ["Flight_ID", "Airline", "Travel_Class", "Average_Ticket_Price"]
    ].head(10)
)

print("Top 10 satisfaction scores:")
display(
    df.sort_values("Passenger_Satisfaction", ascending=False)[
        ["Flight_ID", "Airline", "Passenger_Satisfaction", "Flight_Status"]
    ].head(10)
)

Top 10 flights by delay:


,Flight_ID,Airline,Origin,Destination,Delay_Minutes,Flight_Status
26,FL0027,SpiceJet,Bengaluru,Hyderabad,110.00,Delayed
177,FL0178,IndiGo,Delhi,Mumbai,110.00,Delayed
82,FL0083,Air India Express,Lucknow,Delhi,110.00,Delayed
57,FL0058,IndiGo,Hyderabad,Kolkata,110.00,Delayed
50,FL0051,Air India,Delhi,Mumbai,110.00,Delayed
128,FL0129,SpiceJet,Lucknow,Delhi,110.00,Delayed
146,FL0147,IndiGo,Mumbai,Kolkata,110.00,Delayed
169,FL0170,IndiGo,Mumbai,Kolkata,110.00,Delayed
161,FL0162,Air India,Hyderabad,Kolkata,110.00,Delayed
21,FL0022,Vistara,Kochi,Delhi,75.00,Delayed


Top 10 ticket prices:


,Flight_ID,Airline,Travel_Class,Average_Ticket_Price
138,FL0139,Air India Express,Business,27832
52,FL0053,Akasa Air,Business,21499
32,FL0033,Akasa Air,Business,17625
176,FL0177,Vistara,Business,17306
108,FL0109,SpiceJet,Business,16390
132,FL0133,IndiGo,Business,15324
155,FL0156,Vistara,Business,15040
7,FL0008,Air India,Business,14173
51,FL0052,SpiceJet,Business,13824
92,FL0093,Air India Express,Business,12970


Top 10 satisfaction scores:


,Flight_ID,Airline,Passenger_Satisfaction,Flight_Status
1,FL0002,Akasa Air,5,On Time
13,FL0014,Air India,5,On Time
7,FL0008,Air India,5,On Time
4,FL0005,Air India,5,Delayed
19,FL0020,SpiceJet,5,On Time
14,FL0015,Akasa Air,5,On Time
12,FL0013,SpiceJet,5,On Time
47,FL0048,Air India,5,Delayed
37,FL0038,Air India,5,Cancelled
41,FL0042,IndiGo,5,On Time


## 10. Grouping and Aggregation

### Flights by airline

We compare flight count, passengers, average delay, average ticket price, and satisfaction across airlines.

In [ ]:
airline_summary = (
    df.groupby("Airline")
      .agg(
          Flights=("Flight_ID", "count"),
          Total_Passengers=("Passengers", "sum"),
          Avg_Delay_Minutes=("Delay_Minutes", "mean"),
          Avg_Ticket_Price=("Average_Ticket_Price", "mean"),
          Avg_Satisfaction=("Passenger_Satisfaction", "mean")
      )
      .sort_values("Flights", ascending=False)
)

display(airline_summary.round(2))

,Flights,Total_Passengers,Avg_Delay_Minutes,Avg_Ticket_Price,Avg_Satisfaction
Airline,,,,,
Vistara,37,4484,18.63,"6,265.65",3.38
IndiGo,32,4158,25.03,"5,931.66",3.34
SpiceJet,32,3929,24.14,"6,189.75",3.53
Akasa Air,31,3427,14.40,"5,841.00",3.42
Air India Express,24,3027,11.29,"7,388.46",3.67
Air India,24,3074,21.26,"5,083.08",3.75


### Flight status analysis

In [ ]:
status_summary = (
    df.groupby("Flight_Status")
      .agg(
          Flights=("Flight_ID", "count"),
          Avg_Delay_Minutes=("Delay_Minutes", "mean"),
          Avg_Satisfaction=("Passenger_Satisfaction", "mean")
      )
      .sort_values("Flights", ascending=False)
)

display(status_summary.round(2))

,Flights,Avg_Delay_Minutes,Avg_Satisfaction
Flight_Status,,,
On Time,114,3.46,3.59
Delayed,59,50.00,3.27
Cancelled,7,NaN,3.86


### Weather and delay

In [ ]:
weather_summary = (
    df.groupby("Weather")
      .agg(
          Flights=("Flight_ID", "count"),
          Avg_Delay_Minutes=("Delay_Minutes", "mean"),
          Avg_Satisfaction=("Passenger_Satisfaction", "mean")
      )
      .sort_values("Avg_Delay_Minutes", ascending=False)
)

display(weather_summary.round(2))

,Flights,Avg_Delay_Minutes,Avg_Satisfaction
Weather,,,
Rain,43,22.22,3.37
Storm,30,20.10,3.17
Cloudy,38,20.03,3.84
Clear,34,18.03,3.35
Fog,35,15.79,3.69


### Travel class

In [ ]:
class_summary = (
    df.groupby("Travel_Class")
      .agg(
          Flights=("Flight_ID", "count"),
          Avg_Ticket_Price=("Average_Ticket_Price", "mean"),
          Avg_Passengers=("Passengers", "mean"),
          Avg_Satisfaction=("Passenger_Satisfaction", "mean")
      )
      .sort_values("Avg_Ticket_Price", ascending=False)
)

display(class_summary.round(2))

,Flights,Avg_Ticket_Price,Avg_Passengers,Avg_Satisfaction
Travel_Class,,,,
Business,11,"16,780.00",133.73,3.27
Premium Economy,35,"8,736.74",130.69,3.69
Economy,134,"4,550.27",119.81,3.46


## 11. Data Transformation

New columns are created using Pandas operations:

- **Load_Factor_%** = passengers ÷ seat capacity × 100
- **Estimated_Revenue** = passengers × average ticket price
- **Delay_Category** = categorical classification of delay duration
- **Month** = month extracted from the flight date

In [ ]:
analysis_df = df.copy()

analysis_df["Load_Factor_%"] = (
    analysis_df["Passengers"] / analysis_df["Seat_Capacity"] * 100
).round(2)

analysis_df["Estimated_Revenue"] = (
    analysis_df["Passengers"] * analysis_df["Average_Ticket_Price"]
)

analysis_df["Delay_Category"] = pd.cut(
    analysis_df["Delay_Minutes"],
    bins=[-1, 0, 15, 60, np.inf],
    labels=[
        "No Delay",
        "Minor (1-15 min)",
        "Moderate (16-60 min)",
        "Severe (>60 min)"
    ]
)

analysis_df["Month"] = analysis_df["Flight_Date"].dt.month_name()

display(
    analysis_df[
        [
            "Flight_ID", "Airline", "Passengers", "Seat_Capacity",
            "Load_Factor_%", "Average_Ticket_Price",
            "Estimated_Revenue", "Delay_Minutes", "Delay_Category", "Month"
        ]
    ].head(10)
)

,Flight_ID,Airline,Passengers,Seat_Capacity,Load_Factor_%,Average_Ticket_Price,Estimated_Revenue,Delay_Minutes,Delay_Category,Month
0,FL0001,SpiceJet,210,180,116.67,6894,1447740,50.00,Moderate (16-60 min),March
1,FL0002,Akasa Air,70,210,33.33,4468,312760,0.00,No Delay,June
2,FL0003,Air India,67,220,30.45,4713,315771,5.00,Minor (1-15 min),May
3,FL0004,Akasa Air,153,220,69.55,3791,580023,35.00,Moderate (16-60 min),May
4,FL0005,Air India,178,180,98.89,6505,1157890,20.00,Moderate (16-60 min),February
5,FL0006,Akasa Air,123,220,55.91,5636,693228,35.00,Moderate (16-60 min),May
6,FL0007,Air India Express,178,210,84.76,9430,1678540,0.00,No Delay,June
7,FL0008,Air India,157,220,71.36,14173,2225161,0.00,No Delay,April
8,FL0009,Akasa Air,82,210,39.05,5893,483226,75.00,Severe (>60 min),January
9,FL0010,Akasa Air,125,220,56.82,2500,312500,12.00,Minor (1-15 min),June


## 12. Load Factor Analysis

A load factor above 100% is not operationally realistic for a standard passenger capacity measure. If it occurs in the dataset, it should be treated as a **data quality finding**, not automatically corrected.

In [ ]:
load_factor_summary = (
    analysis_df.groupby("Airline")
    .agg(
        Flights=("Flight_ID", "count"),
        Avg_Load_Factor=("Load_Factor_%", "mean"),
        Max_Load_Factor=("Load_Factor_%", "max")
    )
    .sort_values("Avg_Load_Factor", ascending=False)
)

display(load_factor_summary.round(2))

print("Flights with load factor above 100%:", (analysis_df["Load_Factor_%"] > 100).sum())

,Flights,Avg_Load_Factor,Max_Load_Factor
Airline,,,
IndiGo,32,66.78,110.00
Air India Express,24,64.79,100.00
Air India,24,64.65,108.06
SpiceJet,32,63.60,116.67
Vistara,37,61.87,116.11
Akasa Air,31,55.21,96.19


Flights with load factor above 100%: 9


## 13. Estimated Revenue by Airline

This is a simple analytical estimate based on passengers multiplied by average ticket price. It is **not accounting revenue**, because the dataset does not provide actual transaction level revenue.

In [ ]:
revenue_by_airline = (
    analysis_df.groupby("Airline")
    .agg(
        Estimated_Revenue=("Estimated_Revenue", "sum"),
        Total_Passengers=("Passengers", "sum"),
        Avg_Ticket_Price=("Average_Ticket_Price", "mean")
    )
    .sort_values("Estimated_Revenue", ascending=False)
)

display(revenue_by_airline.round(2))

,Estimated_Revenue,Total_Passengers,Avg_Ticket_Price
Airline,,,
Vistara,28689467,4484,"6,265.65"
IndiGo,25709131,4158,"5,931.66"
SpiceJet,24195935,3929,"6,189.75"
Air India Express,23944164,3027,"7,388.46"
Akasa Air,18872059,3427,"5,841.00"
Air India,16232418,3074,"5,083.08"


## 14. Route-Level Analysis

Routes with frequent flights and routes with high average delays can be identified using multi column grouping.

In [ ]:
route_summary = (
    analysis_df.groupby(["Origin", "Destination"])
    .agg(
        Flights=("Flight_ID", "count"),
        Total_Passengers=("Passengers", "sum"),
        Avg_Delay_Minutes=("Delay_Minutes", "mean"),
        Avg_Satisfaction=("Passenger_Satisfaction", "mean")
    )
    .sort_values("Flights", ascending=False)
)

print("Most frequently operated routes:")
display(route_summary.head(10).round(2))

print("Routes with highest average delay among routes with at least 5 flights:")
display(
    route_summary[route_summary["Flights"] >= 5]
    .sort_values("Avg_Delay_Minutes", ascending=False)
    .head(10)
    .round(2)
)

Most frequently operated routes:


Flights  Total_Passengers  Avg_Delay_Minutes  \
Origin    Destination                                                 
Delhi     Chennai           17              2011              19.00   
Mumbai    Bengaluru         16              1933              15.44   
Srinagar  Delhi             15              1957              16.36   
Kochi     Delhi             14              1657              15.00   
Delhi     Srinagar          13              1828              18.92   
          Mumbai            13              1609              39.18   
Bengaluru Hyderabad         13              1795              15.38   
Mumbai    Kolkata           13              1683              27.00   
Ahmedabad Bengaluru         12              1411              11.45   
Lucknow   Delhi             11              1137              40.00   

                       Avg_Satisfaction  
Origin    Destination                    
Delhi     Chennai                  3.47  
Mumbai    Bengaluru                3.25  
Srinagar  Delhi                    3.27  
Kochi     Delhi                    3.50  
Delhi     Srinagar                 3.31  
          Mumbai                   3.54  
Bengaluru Hyderabad                3.77  
Mumbai    Kolkata                  3.92  
Ahmedabad Bengaluru                3.92  
Lucknow   Delhi                    3.18

Routes with highest average delay among routes with at least 5 flights:


Flights  Total_Passengers  Avg_Delay_Minutes  \
Origin    Destination                                                 
Hyderabad Kolkata            7               884              51.71   
Lucknow   Delhi             11              1137              40.00   
Delhi     Mumbai            13              1609              39.18   
Mumbai    Kolkata           13              1683              27.00   
Delhi     Chennai           17              2011              19.00   
          Srinagar          13              1828              18.92   
Srinagar  Delhi             15              1957              16.36   
Mumbai    Bengaluru         16              1933              15.44   
Bengaluru Hyderabad         13              1795              15.38   
Kochi     Delhi             14              1657              15.00   

                       Avg_Satisfaction  
Origin    Destination                    
Hyderabad Kolkata                  3.86  
Lucknow   Delhi                    3.18  
Delhi     Mumbai                   3.54  
Mumbai    Kolkata                  3.92  
Delhi     Chennai                  3.47  
          Srinagar                 3.31  
Srinagar  Delhi                    3.27  
Mumbai    Bengaluru                3.25  
Bengaluru Hyderabad                3.77  
Kochi     Delhi                    3.50

## 15. Monthly Analysis

The transformed `Month` column allows us to compare operations and customer satisfaction across months.

In [ ]:
monthly_summary = (
    analysis_df.groupby("Month", sort=False)
    .agg(
        Flights=("Flight_ID", "count"),
        Avg_Delay_Minutes=("Delay_Minutes", "mean"),
        Avg_Satisfaction=("Passenger_Satisfaction", "mean"),
        Total_Passengers=("Passengers", "sum")
    )
)

month_order = [
    "January", "February", "March",
    "April", "May", "June"
]

monthly_summary = monthly_summary.reindex(month_order)
display(monthly_summary.round(2))

,Flights,Avg_Delay_Minutes,Avg_Satisfaction,Total_Passengers
Month,,,,
January,29,22.10,3.45,3656
February,28,22.75,3.89,3502
March,33,10.69,3.67,3839
April,35,25.41,3.26,4354
May,30,16.93,3.03,3504
June,25,16.76,3.76,3244


## 16. Booking Channel Analysis

In [ ]:
booking_summary = (
    analysis_df.groupby("Booking_Channel")
    .agg(
        Flights=("Flight_ID", "count"),
        Avg_Ticket_Price=("Average_Ticket_Price", "mean"),
        Avg_Satisfaction=("Passenger_Satisfaction", "mean")
    )
    .sort_values("Flights", ascending=False)
)

display(booking_summary.round(2))

,Flights,Avg_Ticket_Price,Avg_Satisfaction
Booking_Channel,,,
Online Travel Portal,50,"5,636.28",3.74
Mobile App,45,"6,637.24",3.44
Travel Agency,44,"6,223.89",3.39
Airline Website,41,"5,994.17",3.37


## 17. Overall KPI Snapshot

In [ ]:
kpis = pd.Series({
    "Total Flights": len(analysis_df),
    "Total Passengers": analysis_df["Passengers"].sum(),
    "On-Time Flights": (analysis_df["Flight_Status"] == "On Time").sum(),
    "Delayed Flights": (analysis_df["Flight_Status"] == "Delayed").sum(),
    "Cancelled Flights": (analysis_df["Flight_Status"] == "Cancelled").sum(),
    "Average Delay (min)": analysis_df["Delay_Minutes"].mean(),
    "Average Satisfaction": analysis_df["Passenger_Satisfaction"].mean(),
    "Average Load Factor (%)": analysis_df["Load_Factor_%"].mean(),
    "Estimated Revenue": analysis_df["Estimated_Revenue"].sum()
})

display(kpis.to_frame("Value"))

,Value
Total Flights,180.00
Total Passengers,"22,099.00"
On-Time Flights,114.00
Delayed Flights,59.00
Cancelled Flights,7.00
Average Delay (min),19.34
Average Satisfaction,3.49
Average Load Factor (%),62.66
Estimated Revenue,"137,643,174.00"


## 18. Key Observations

Based on the analysis of the supplied dataset:

1. **On time performance dominates:** 114 of 180 flights are marked On Time, while 59 are Delayed and 7 are Cancelled.
2. **Delays affect satisfaction:** delayed flights have an average passenger satisfaction of about **3.27/5**, compared with about **3.59/5** for on time flights.
3. **Rain is associated with the highest average delay:** among the weather categories, Rain has an average delay of about **22.22 minutes**, followed by Storm at about 20.10 minutes.
4. **Vistara has the highest estimated revenue:** using the simple passengers × average ticket price calculation, Vistara generates the largest estimated revenue in this dataset.
5. **Premium Economy has the highest satisfaction among travel classes:** its average satisfaction is about **3.69/5**, higher than Economy and Business.
6. **April shows the highest average delay:** April has an average delay of about **25.41 minutes**, while March has the lowest at about **10.69 minutes**.
7. **There are missing delay values:** all 7 missing `Delay_Minutes` records correspond to Cancelled flights. These should not automatically be replaced with zero because a cancelled flight does not necessarily mean zero delay.
8. **The dataset contains load factor anomalies:** several flights have more passengers than recorded seat capacity, producing load factors above 100%. This is an important data quality issue that should be investigated before using load factor for operational decisions.

## 19. Conclusion

The Pandas workflow shows how a raw flight operations dataset can be converted into useful operational insights. The analysis covered data loading, inspection, descriptive statistics, selection, filtering, sorting, grouping, aggregation, and transformation.

The strongest patterns are related to **flight status, delays, weather, travel class, monthly performance, revenue estimates, and data quality**. The analysis also demonstrates why missing values and unrealistic derived metrics should be investigated before drawing business conclusions.